storm-core/src/jvm/backtype/storm/metric/internal/CountStatAndMetric.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Constructor
+     * @param numBuckets the number of buckets to divide the time periods into.
+     * @param startTime if positive the simulated time to start the from.
+     */
+    CountStatAndMetric(int numBuckets, long startTime){
+        numBuckets = Math.max(numBuckets, 2);
+        //We want to capture the full time range, so the target size is as
+        // if we had one bucket less, then we do
+        _tmSize = 10 * 60 * 1000 / (numBuckets - 1);
+        _thSize = 3 * 60 * 60 * 1000 / (numBuckets - 1);
+        _odSize = 24 * 60 * 60 * 1000 / (numBuckets - 1);
+        if (_tmSize < 1 || _thSize < 1 || _odSize < 1) {
+            throw new IllegalArgumentException("number of buckets is too large to be supported");
+        }
+        _tmBuckets = new long[numBuckets];
+        _tmTime = new long[numBuckets];
+        _thBuckets = new long[numBuckets];
+        _thTime = new long[numBuckets];
+        _odBuckets = new long[numBuckets];
+        _odTime = new long[numBuckets];
+        _allTime = 0;
+        _exactExtra = 0;
+
+        _bucketStart = startTime >= 0 ? startTime : System.currentTimeMillis();
+        _currentBucket = new AtomicLong(0);
+        if (startTime < 0) {
+            _task = new Fresher();
+            MetricStatTimer._timer.scheduleAtFixedRate(_task, _tmSize, _tmSize);
+        } else {
+            _task = null;
+        }
+    }
+
+    /**
+     * Increase the count by the given value.
+     *
+     * @param count number to count
+     */
+    public void incBy(long count) {
+        _currentBucket.addAndGet(count);
+    }
+
+   
+
+    @Override
+    public synchronized Object getValueAndReset() {
+        return getValueAndReset(System.currentTimeMillis());
+    }
+
+    synchronized Object getValueAndReset(long now) {
+        long value = _currentBucket.getAndSet(0);
+        long timeSpent = now - _bucketStart;
+        long ret = value + _exactExtra;
+        _bucketStart = now;
+        _exactExtra = 0;
+        rotateBuckets(value, timeSpent);
+        return ret;
+    }
+
+    synchronized void rotateSched(long now) {
+        long value = _currentBucket.getAndSet(0);
+        long timeSpent = now - _bucketStart;
+        _exactExtra += value;
+        _bucketStart = now;
+        rotateBuckets(value, timeSpent);
+    }
+
+    synchronized void rotateBuckets(long value, long timeSpent) {
+        rotate(value, timeSpent, _tmSize, _tmTime, _tmBuckets);
+        rotate(value, timeSpent, _thSize, _thTime, _thBuckets);
+        rotate(value, timeSpent, _odSize, _odTime, _odBuckets);
+        _allTime += value;
+    }
+
+    private synchronized void rotate(long value, long timeSpent, long targetSize, long [] times, long [] buckets) {
+        times[0] += timeSpent;
+        buckets[0] += value;
+
+        long currentTime = 0;
+        long currentVal = 0;
+        if (times[0] >= targetSize) {
+            for (int i = 0; i < buckets.length; i++) {
```

Comment:
Yes we could, but I don't think it is going to make that much of a difference.  I purposely moved rotating the metrics off of the critical path.  The call to rotate happens in a background thread, that is shared by all similar metrics in a worker process.  This allows the critical path, the thing that happens all of the time, to be just an atomic increment.  Rotating only happens every 30 seconds or so too.  So I suspect the best that it is going to do is reduce the overall CPU the worker uses by a tiny amount.